src/service/impl/departures/index.ts

```diff
@@ -57,7 +71,115 @@ export default (
     pubOpts
   );
 
+  const fetchDeparturesForLine = async (
+    quayId: string,
+    lineId: string,
+    lineName: string | undefined
+  ) => {
+    try {
+      const result = await journeyPlannerClient_v3.query<
+        FavouriteDepartureQuery,
+        FavouriteDepartureQueryVariables
+      >({
+        query: FavouriteDepartureDocument,
+        variables: {
+          quayIds: quayId,
+          lines: lineId
+        }
+      });
+
+      if (result.errors) {
+        return Result.err(new APIError(result.errors));
+      }
+      return Result.ok(result.data);
+    } catch (error) {
+      return Result.err(new APIError(error));
+    }
+  };
+
   const api: IDeparturesService = {
+    async getFavouriteDepartures(query) {
+      return Promise.all(
+        query.map(async fav => {
+          return await fetchDeparturesForLine(
+            fav.quayId,
+            fav.lineId,
+            fav.lineName
+          );
+        })
+      )
+        .then(results => {
+          const validResults = results
+            .filter(res => res.isOk)
+            .map(res => res.unwrap());
+
+          // Create Departure Groups
+          const lineInfos = extractLineInfos(validResults);
+          const departureGroups: DepartureGroup[] = lineInfos.map(lineInfo => {
+            return {
+              lineInfo: lineInfo,
+              departures: []
+            };
+          });
+
+          // Add calls to DepartureGroups
+          validResults
+            .map(result => result.quays)
+            .flatMap(quay => quay)
+            .map(quay => {
+              return quay.estimatedCalls;
+            })
+            .flatMap(call => call)
+            .filter(call => callMatchesQueriedLineName(call, query))
+            .forEach(call => {
+              departureGroups
+                .find(group => {
+                  return (
+                    group.lineInfo?.lineId === call.serviceJourney?.line.id &&
+                    group.lineInfo?.lineName ===
+                      call.destinationDisplay?.frontText
+                  );
+                })
+                ?.departures.push({
+                  aimedTime: call.aimedDepartureTime,
+                  serviceDate: call.date,
+                  time: call.expectedDepartureTime,
+                  situations: [],
+                  serviceJourneyId: call.serviceJourney?.id
+                });
+            });
```

Comment:
That's a lot of mapping on one line 🤯 Would be good to talk trough this (and the rest of this file).